tutorials/v7/ntuple/ntpl007_mtFill.C

```diff
@@ -0,0 +1,153 @@
+/// \file
+/// \ingroup tutorial_ntuple
+/// \notebook
+/// Example of multi-threaded writes using multuple REntry objects
+///
+/// \macro_image
+/// \macro_code
+///
+/// \date July 2021
+/// \author The ROOT Team
+
+// NOTE: The RNTuple classes are experimental at this point.
+// Functionality, interface, and data format is still subject to changes.
+// Do not use for real data!
+
+// Until C++ runtime modules are universally used, we explicitly load the ntuple library.  Otherwise
+// triggering autoloading from the use of templated types would require an exhaustive enumeration
+// of "all" template instances in the LinkDef file.
+R__LOAD_LIBRARY(ROOTNTuple)
+
+#include <ROOT/RNTuple.hxx>
+#include <ROOT/RNTupleModel.hxx>
+
+#include <TCanvas.h>
+#include <TH1F.h>
+#include <TRandom.h>
+#include <TSystem.h>
+
+#include <atomic>
+#include <memory>
+#include <mutex>
+#include <thread>
+#include <vector>
+#include <utility>
+
+// Import classes from experimental namespace for the time being
+using REntry = ROOT::Experimental::REntry;
+using RNTupleModel = ROOT::Experimental::RNTupleModel;
+using RNTupleReader = ROOT::Experimental::RNTupleReader;
+using RNTupleWriter = ROOT::Experimental::RNTupleWriter;
+
+// Where to store the ntuple of this example
+constexpr char const *kNTupleFileName = "ntpl007_mtFill.root";
+
+// Number of parallel threads to fill the ntuple
+constexpr int kNWriterThreads = 4;
+
+// Number of events to generate is kNEventsPerThread * kNWriterThreads
+constexpr int kNEventsPerThread = 25000;
+
+// Thread function to generate and write events
+void FillData(std::unique_ptr<REntry> entry, RNTupleWriter *ntuple) {
+   // Protect the ntuple->Fill() call
+   static std::mutex gLock;
+
+   static std::atomic<std::uint32_t> gThreadId;
+   auto threadId = gThreadId++;
+
+   auto prng = std::make_unique<TRandom3>();
+   prng->SetSeed();
+
+   auto id = entry->Get<std::uint32_t>("id");
+   auto vpx = entry->Get<std::vector<float>>("vpx");
+   auto vpy = entry->Get<std::vector<float>>("vpy");
+   auto vpz = entry->Get<std::vector<float>>("vpz");
+
+   for (int i = 0; i < kNEventsPerThread; i++) {
+      vpx->clear();
+      vpy->clear();
+      vpz->clear();
+      *id = threadId;
+
+      int npx = static_cast<int>(prng->Rndm(1) * 15);
+      // Set the field data for the current event
+      for (int j = 0; j < npx; ++j) {
+         float px, py, pz;
+         prng->Rannor(px, py);
+         pz = px*px + py*py;
+
+         vpx->emplace_back(px);
+         vpy->emplace_back(py);
+         vpz->emplace_back(pz);
+      }
+
+      std::lock_guard<std::mutex> guard(gLock);
```

Comment:
As a side questions, is this (user locking call to Fill) the final interface or will the lock/use-of-atomics get added to RNTuple later?

Reply:
I actually have the same question. I would prefer the lock is internal, but I imagine we would need a non synchronized version for the single thread case.

A was also thinking about a non blocking version of `Fill` (e.g. `RNTuple::Writer(std::unique_ptr<REntry>)` that would enqueue the serialization for a dedicated thread. Of course you have to pay for the allocation/deallocation of `REntry` at every event:
```cpp
for(auto& event: all_events) {
  auto entry = ntuple->CreateEntry();
  auto &vpx = *entry->Get<std::vector<float>>("vpx");
  vpx.clear();
  vpx.push_back(42.);
  ntuple->Fill(std::move(entry));
}
```

Reply:
I was planning to keep `Fill()` unlocked for consistency reasons. While we could protect `Fill()`, the rest of the `RNTupleWriter` would still not be prepared to be called concurrently. I think it is cleaner to keep the class [conditionally thread safe](https://root.cern/manual/thread_safety/), like most other RNTuple classes.

Moving the REntries into a serialization queue is a really interesting idea! I don't think we need/should to address this in the `RNTupleWriter` class though. What we essentially need is a bounded, thread-safe queue for the producer-consumer thread synchronization task. I wonder if an implementation of such a queue would make sense as a generally available utility class in ROOT, @pcanal?

That said, using a dedicated thread for the serialization would only make sense if we actually spend a non-negligible amount of time in copying data from the in-memory objects into the I/O buffers.  For compressing I/O buffers, we already have infrastructure in place to do it multi-threaded.  That helps substantially for simple events but it is, as far as I can see, not the main bottleneck for writing/creating complex events.

@pikacic Perhaps we can discuss it in person?

Reply:
"bounded, thread-safe queue for the producer-consumer thread synchronization task": yes, it does sound useful.  The challenging part is the "bounded" part and more particularly what to do when the bound is reached.

Reply:
I have in mind an implementation with conditional variables, so enqueuing would block if the queue is full.

Reply:
In some context (CMSSW for example), they have argued that this is sub-optiomal as it prevents the calling thread from doing other work while waiting and in similar context (eg TFileBufferMerger, we have to provide feedback on how full the queue was, some of the discussion also talked about a call back of sort). [@Dr15Jones @dan131riley might have more details]

Reply:
So it looks to me that in the example, the most time consuming call would actually be `tuple->Fill(*entry);`? If so, then you would get very little concurrency. If the `Fill` call could make use of available threads (much like `TTree::Fill` if ImplicitMT is on) then you'd still have a problem since you'd have lots of possible threads all being blocked and therefore unusable by the `Fill` call. That would mean you'd have to oversubcribed the system, which isn't really a good idea.

An alternative design used frequently in CMSSW is to have a `SerialTaskQueue` with an associated callback for when the task finishes. The callback is used to asynchronously start another series of tasks. This is often done recursively: an asynchronous function setup work, when the work completes it enqueues a finishing task and then when the finishing task runs its last step is to call the asynchronous function which had started the whole process.

Below is a rewrite of the `Write()` function using this idea (it depends on code similar to https://github.com/cms-sw/cmssw/blob/master/FWCore/Concurrency/interface/SerialTaskQueue.h) using tbb::task_group to handle the thread pool. In the code, no TBB threads are ever blocked so they are always available for work (e.g. by `Fill`).

```C++
void nextEntryAsync(tbb::task_group& group, std::atomic<int>& countDown, int threadID, REntry& entry, RNTupleWriter& tuple, SerialTaskQueue& queue, TRandom3&rand) {

	group.run([]() {
		if(--countDown >=0) {
      auto id = entry->Get<std::uint32_t>("id");
      auto vpx = entry->Get<std::vector<float>>("vpx");
      auto vpy = entry->Get<std::vector<float>>("vpy");
      auto vpz = entry->Get<std::vector<float>>("vpz");
			
      vpx->clear();
      vpy->clear();
      vpz->clear();
      *id = threadID;
      
      int npx = static_cast<int>(prng->Rndm(1) * 15);
      // Set the field data for the current event
      vpx->reserve(npx);
      vpy->reserve(npy);
      vpz->reserve(npz);
      for (int j = 0; j < npx; ++j) {
          float px, py, pz;
          prng->Rannor(px, py);
          pz = px*px + py*py;

          vpx->emplace_back(px);
          vpy->emplace_back(py);
          vpz->emplace_back(pz);
      }      
      
      //only one thread at a time can fill the tuple
      // once fill is done, try to process next entry
      queue.push(group, [&,threadID]() {
        //internally, Fill could use available TBB threads
        tuple.Fill(entry);
        nextEntryAsync(group, countDown, threadID, entry, tuple, queue, rand);
      });
		}
	});
}

void Write() {
  
  // Create the data model
  auto model = RNTupleModel::Create();
  model->MakeField<std::uint32_t>("id");
  model->MakeField<std::vector<float>>("vpx");
  model->MakeField<std::vector<float>>("vpy");
  model->MakeField<std::vector<float>>("vpz");

  // We hand-over the data model to a newly created ntuple of name "NTuple", stored in kNTupleFileName
  auto ntuple = RNTupleWriter::Recreate(std::move(model), "NTuple", kNTupleFileName);
  
  //set the number of threads total as well as the number of threads for the local arena
  // the global setting makes sure any other arenas used will have their max threads restricted
  tbb::global_control control(tbb::global_control::max_allowed_parallelism, nWriterThreads);
  tbb::task_arena arena(nWriterThreads);
  arena.execute([&]() {
  
    std::atomic<int> countDown = kNEvents;
    
    SerialTaskQueue queue;
    
    std::vector<std::unique_ptr<REntry>> entries;
    std::vector<std::unique_ptr<TRandom3>> rands;
    entries.reserve(nWriterThreads);
    rands.reserve(nWriterThreads);
    
    tbb::task_group group;
    for(int i=0; i< nWriterThreads; ++i) {
      entries.emplace_back(ntuple->CreateEntry());
      rands.emplace_back(std::make_unique<TRandom3>());
      rands.back()->SetSeed();
      nextEntryAsync(group, countDown, i, *entries.back(), ntuple, queue, *rands.back());
    }
    group.wait();
  });
}
```

Reply:
One item of note in this design, the SerialTaskQueue is implicitly bound because at most nWriteThreads-1 tasks can be waiting in the queue at any one time. That is because new tasks can only be spawned after a task has finished in the queue.

Reply:
So pushing things a bit further using the 'task chaining' code (seen here https://github.com/cms-sw/cmssw/blob/master/FWCore/Concurrency/test/test_catch2_WaitingTaskChain.cc) the part of `Write()` that falls within the `arena.execute(...)` could probably be written as

```C++
std::atomic<int> countDown = 0;

SerialTaskQueue queue;

//setup per 'lane' data structures
std::vector<std::unique_ptr<REntry>> entries;
std::vector<std::unique_ptr<TRandom3>> rands;
entries.reserve(nWriterThreads);
rands.reserve(nWriterThreads);

tbb::task_group group;
FinalWaitingTask finalTask;
{
  //this ensures the finalTask can't be run until after the loop finishes
  WaitingTaskHolder final(group, finalTask);

  //create as many 'lanes' as threads
  for(int i=0; i< nWriterThreads; ++i)) {
    entries.emplace_back(ntuple->CreateEntry());
    auto entry = entries.back().get();
    rands.emplace_back(std::make_unique<TRandom3>());
    auto rand = rands.back().get();
    rand->SetSeed();
    
    //here is the 'chain' of tasks to be exetued
    chain::while([&count, kNEvents](){return ++count< kNEvents;},
        chain::first([&](auto iTask){ fillEntry(*entry); } ) 
        | chain::then([&](auto iTask){ queue.push([iTask]() { tuple->Fill(*entry); })})
    )   
    | runLast(final);
  }
}

group.wait();
```

The `chain::while` ability doesn't exist in the present code, but I believe it would be possible to create.